ha/ha.go

```diff
@@ -4,181 +4,204 @@ package ha
+		"REPLACE INTO icingadb_instance(id, environment_id, endpoint_id, responsible, heartbeat, startup_id,"+
 			" icinga2_version, icinga2_start_time, icinga2_notifications_enabled,"+
 			" icinga2_active_service_checks_enabled, icinga2_active_host_checks_enabled,"+
 			" icinga2_event_handlers_enabled, icinga2_flap_detection_enabled,"+
-			" icinga2_performance_data_enabled) VALUES (?, ?, ?, ?, 'y', ?, ?, ?, ?, ?, ?, ?, ?)",
-		h.uid[:],
-		h.super.EnvId,
-		env.Icinga2.EndpointId,
-		h.lastHeartbeat,
-		env.Icinga2.Version,
-		int64(env.Icinga2.ProgramStart*1000),
-		utils.Bool[env.Icinga2.NotificationsEnabled],
-		utils.Bool[env.Icinga2.ActiveServiceChecksEnabled],
-		utils.Bool[env.Icinga2.ActiveHostChecksEnabled],
-		utils.Bool[env.Icinga2.EventHandlersEnabled],
-		utils.Bool[env.Icinga2.FlapDetectionEnabled],
-		utils.Bool[env.Icinga2.PerformanceDataEnabled],
+			" icinga2_performance_data_enabled) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
+		h.instanceId[:],                                    // id
+		h.super.EnvId,                                      // environment_id
+		env.Icinga2.EndpointId,                             // endpoint_id
+		utils.Bool[isActive],                               // responsible
+		h.lastHeartbeat,                                    // heartbeat
+		h.startupId[:],                                     // startup_id
+		env.Icinga2.Version,                                // icinga2_version
+		int64(env.Icinga2.ProgramStart*1000),               // icinga2_start_time
+		utils.Bool[env.Icinga2.NotificationsEnabled],       // icinga2_notifications_enabled
+		utils.Bool[env.Icinga2.ActiveServiceChecksEnabled], // icinga2_active_service_checks_enabled
+		utils.Bool[env.Icinga2.ActiveHostChecksEnabled],    // icinga2_active_host_checks_enabled
+		utils.Bool[env.Icinga2.EventHandlersEnabled],       // icinga2_event_handlers_enabled
+		utils.Bool[env.Icinga2.FlapDetectionEnabled],       // icinga2_flap_detection_enabled
+		utils.Bool[env.Icinga2.PerformanceDataEnabled],     // icinga2_performance_data_enabled
 	)
 	return err
 }
 
-func (h *HA) getInstance() (bool, uuid.UUID, int64, error) {
-	rows, err := h.super.Dbw.SqlFetchAll(
-		mysqlObservers.selectIdHeartbeatFromIcingadbInstanceByEnvironmentId,
-		"SELECT id, heartbeat from icingadb_instance where environment_id = ? LIMIT 1",
-		h.super.EnvId,
+func (h *HA) getActiveInstance(tx connection.DbTransaction) (bool, []byte, error) {
+	rows, err := h.super.Dbw.SqlFetchAllTx(
+		tx, mysqlObservers.selectIdHeartbeatResponsibleFromIcingadbInstanceByEnvironmentId,
+		"SELECT id, heartbeat FROM icingadb_instance"+
+			" WHERE environment_id = ? AND responsible = ? AND heartbeat > ?",
+		h.super.EnvId, utils.Bool[true], utils.TimeToMillisecs(time.Now())-heartbeatTimeoutMillisecs,
 	)
-
 	if err != nil {
-		return false, uuid.UUID{}, 0, err
+		return false, nil, err
 	}
+	if len(rows) > 1 {
+		return false, nil, errors.New("there is more than one active IcingaDB instance")
```

Comment:
Just being defensive here. This should never happen unless someone messes with the database. But won't hurt to detect it then instead of having some weird behavior.

Reply:
Fine.